Add navigator single thread dispatcher scope to ThreadController

Guardiola31337 opened the pull request:
## Description

Adds navigator single thread dispatcher scope to `ThreadController` use it in `MapboxTripSession` for NN interactions instead of `IO` removing the need of the `MapboxNativeNavigatorImpl` `Mutex`

Follow up from https://github.com/mapbox/mapbox-navigation-android/pull/3424#issuecomment-672375626

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Have a single thread dispatcher dedicated for NN so that the `Mutex` from `MapboxNativeNavigatorImpl` can be removed

### Implementation

- Add navigator single thread dispatcher scope to `ThreadController` and use it in `MapboxTripSession` for NN interactions instead of `IO`
- Remove `MapboxNativeNavigatorImpl` `Mutex`

## Testing

We should run extensive and field testing but from the local one everything looks to work 👌 

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

kmadsen commented on libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ThreadController.kt:
```diff
@@ -47,30 +47,37 @@ fun Exception.ifChannelException(action: () -> Unit) {
     val IODispatcher: CoroutineDispatcher =
         Executors.newFixedThreadPool(maxCoresUsed).asCoroutineDispatcher()
+    internal val navigatorDispatcher: CoroutineDispatcher =
+        Executors.newFixedThreadPool(SINGLE_THREAD).asCoroutineDispatcher()
```
+1

kmadsen commented on libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ThreadController.kt:
```diff
@@ -103,6 +110,14 @@ object ThreadController {
     }
 
+    /**
+     * Same as [getIOScopeAndRootJob], but using the Navigator dispatcher.
```
Can we create documentation here independent of the IO scope

"Use this [JobControl] whenever interacting with the navigator.."

kmadsen commented on libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ThreadController.kt:
```diff
@@ -47,30 +47,37 @@ fun Exception.ifChannelException(action: () -> Unit) {
     )
     val IODispatcher: CoroutineDispatcher =
         Executors.newFixedThreadPool(maxCoresUsed).asCoroutineDispatcher()
+    internal val navigatorDispatcher: CoroutineDispatcher =
```
nit: throw a capital in there since it is a constant. appears more consistent with IODispatcher

```suggestion
    internal val NavigatorDispatcher: CoroutineDispatcher =
```

kmadsen commented on libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:
```diff
@@ -75,9 +72,7 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
      * @return true if the raw location was usable, false if not.
      */
     override suspend fun updateLocation(rawLocation: Location, date: Date): Boolean {
```
Can we move the dispatcher definition near MapboxNativeNavigatorImpl

It would be nice to be able to define each of these functions with something like `withContext(NativeDispatcher)`

korshaknn replied:
```suggestion
    override fun updateLocation(rawLocation: Location, date: Date): Boolean {
```

Guardiola31337 commented:
Addressed the NITs. This is ready for another round of 👀 

Also @kmadsen could you cut a follow up ticket to discuss https://github.com/mapbox/mapbox-navigation-android/pull/3457#discussion_r472517285 when you have a chance? Would love to sync up first so we're on the same page.

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -434,7 +434,7 @@ internal class MapboxTripSession(
 
     private suspend fun getNavigatorStatus(date: Date): TripStatus =
-        withContext(ioJobController.scope.coroutineContext) {
+        withContext(navigatorJobController.scope.coroutineContext) {
```
```suggestion
        withContext(NavigatorDispatcher) {
```

korshaknn commented on libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:
```diff
@@ -106,15 +101,13 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
      * is earlier than a previous call, the last status will be returned. The function does not support re-winding time.
      */
     override suspend fun getStatus(date: Date): TripStatus {
```
```suggestion
    override fun getStatus(date: Date): TripStatus {
```

korshaknn commented on libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:
```diff
@@ -134,14 +127,12 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
         route: DirectionsRoute?,
         legIndex: Int
```
remove `suspend` modifier

korshaknn commented on libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ThreadController.kt:
```diff
@@ -47,30 +47,37 @@ fun Exception.ifChannelException(action: () -> Unit) {
     )
     val IODispatcher: CoroutineDispatcher =
         Executors.newFixedThreadPool(maxCoresUsed).asCoroutineDispatcher()
+    internal val NavigatorDispatcher: CoroutineDispatcher =
```
```suggestion
   val NavigatorDispatcher: CoroutineDispatcher =
```
it should be possible to call `withContext(NavigatorDispatcher)`

Guardiola31337 commented:
Addressed your comments @korshaknn Ready for another round 💃

korshaknn reviewed (approved):
LGTM!

Guardiola31337 commented:
🆗 going ahead and merging here. Will follow up with https://github.com/mapbox/mapbox-navigation-android/issues/3463 in a separate PR.
